File: src/main/java/plus/maa/backend/common/aop/JsonSchemaAop.java
package plus.maa.backend.common.aop;


import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Before;
import org.aspectj.lang.annotation.Pointcut;
import org.everit.json.schema.Schema;
import org.everit.json.schema.ValidationException;
import org.everit.json.schema.loader.SchemaLoader;
import org.json.JSONObject;
import org.json.JSONTokener;
import org.springframework.core.io.ClassPathResource;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import plus.maa.backend.common.annotation.JsonSchema;
import plus.maa.backend.controller.request.comments.CommentsRatingDTO;
import plus.maa.backend.controller.request.copilot.CopilotCUDRequest;
import plus.maa.backend.controller.request.copilot.CopilotRatingReq;
import plus.maa.backend.controller.response.MaaResultException;

import java.io.IOException;
import java.io.InputStream;


/**
 * @author LoMu
 * Date  2023-01-22 17:53
 */

@Component
@Aspect
@Slf4j
@RequiredArgsConstructor
public class JsonSchemaAop {
    private final ObjectMapper mapper;
    private static final String COPILOT_SCHEMA_JSON = "static/templates/maa-copilot-schema.json";
    private static final String RATING_SCHEMA_JSON = "static/templates/maa-rating-schema.json";

    @Pointcut("@annotation(plus.maa.backend.common.annotation.JsonSchema)")
    public void pt() {
    }

    /**
     * 数据校验
     *
     * @param joinPoint  形参
     * @param jsonSchema 注解
     */
    @Before("pt() && @annotation(jsonSchema)")
    public void before(JoinPoint joinPoint, JsonSchema jsonSchema) {
        String schema_json = null;
        String content = null;
        //获取 CopilotCUDRequest形参的index
        for (Object arg : joinPoint.getArgs()) {
            if (arg instanceof CopilotCUDRequest) {
                content = ((CopilotCUDRequest) arg).getContent();
                schema_json = COPILOT_SCHEMA_JSON;
            }
            if (arg instanceof CopilotRatingReq || arg instanceof CommentsRatingDTO) {
                try {
                    schema_json = RATING_SCHEMA_JSON;
                    content = mapper.writeValueAsString(arg);
                } catch (JsonProcessingException e) {
                    log.error("json解析失败", e);
                }
            }
        }
        if (content == null) return;


        //获取json schema json路径并验证
        try (InputStream inputStream = new ClassPathResource(schema_json).getInputStream()) {
            JSONObject json = new JSONObject(content);
            JSONObject jsonObject = new JSONObject(new JSONTokener(inputStream));
            Schema schema = SchemaLoader.load(jsonObject);
            schema.validate(json);
        } catch (IOException e) {
            throw new RuntimeException(e);
        } catch (ValidationException e) {
            throw new MaaResultException(HttpStatus.BAD_REQUEST.value(), "数据不符合规范，请前往前端作业编辑器进行操作");
        }
    }
}

File: src/main/kotlin/plus/maa/backend/common/aop/JsonSchemaAop.kt
package plus.maa.backend.common.aop

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import io.github.oshai.kotlinlogging.KotlinLogging
import org.aspectj.lang.JoinPoint
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Before
import org.aspectj.lang.annotation.Pointcut
import org.everit.json.schema.ValidationException
import org.everit.json.schema.loader.SchemaLoader
import org.json.JSONObject
import org.json.JSONTokener
import org.springframework.core.io.ClassPathResource
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Component
import plus.maa.backend.common.annotation.JsonSchema
import plus.maa.backend.controller.request.comments.CommentsRatingDTO
import plus.maa.backend.controller.request.copilot.CopilotCUDRequest
import plus.maa.backend.controller.request.copilot.CopilotRatingReq
import plus.maa.backend.controller.response.MaaResultException
import java.io.IOException

private val log = KotlinLogging.logger { }

/**
 * @author LoMu
 * Date  2023-01-22 17:53
 */
@Component
@Aspect
class JsonSchemaAop(
    private val mapper: ObjectMapper
) {
    @Pointcut("@annotation(plus.maa.backend.common.annotation.JsonSchema)")
    fun pt() {
    }

    /**
     * 数据校验
     *
     * @param joinPoint  形参
     * @param jsonSchema 注解
     */
    @Before("pt() && @annotation(jsonSchema)")
    fun before(joinPoint: JoinPoint, jsonSchema: JsonSchema?) {
        var schemaJson: String? = null
        var content: String? = null
        //判断是验证的是Copilot还是Rating
        for (arg in joinPoint.args) {
            if (arg is CopilotCUDRequest) {
                content = arg.content
                schemaJson = COPILOT_SCHEMA_JSON
            }
            if (arg is CopilotRatingReq || arg is CommentsRatingDTO) {
                try {
                    schemaJson = RATING_SCHEMA_JSON
                    content = mapper.writeValueAsString(arg)
                } catch (e: JsonProcessingException) {
                    log.error(e) { "json解析失败" }
                }
            }
        }
        if (schemaJson == null || content == null) return


        //获取json schema json路径并验证
        try {
            ClassPathResource(schemaJson).inputStream.use { inputStream ->
                val json = JSONObject(content)
                val jsonObject = JSONObject(JSONTokener(inputStream))
                val schema = SchemaLoader.load(jsonObject)
                schema.validate(json)
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        } catch (e: ValidationException) {
            log.warn { "schema Location: ${e.violatedSchema.schemaLocation}" }
            throw MaaResultException(HttpStatus.BAD_REQUEST.value(), "数据不符合规范，请前往前端作业编辑器进行操作")
        }
    }

    companion object {
        private const val COPILOT_SCHEMA_JSON = "static/templates/maa-copilot-schema.json"
        private const val RATING_SCHEMA_JSON = "static/templates/maa-rating-schema.json"
    }
}